### PR TITLE
Added push-to-external application feature documentation

### LINKED ISSUE #207: Push to external application
There is no documentation about configuring push-to-external-application nor the actual use by clicking in the toolbar, in the menu and also not on selecting a different external application to push to. See [JabRef/jabref#4991](https://github.com/JabRef/jabref/pull/4991)

### PR DESCRIPTION
**Pull request by michizhou**
Added new documentation with details about usage of the push-to-external-application feature through the tools submenu and configuration of the feature that allows one to select which external application a user wishes to push to. Also modified heading to comply with markdown lint rules.

Closes #207

### CODE REVIEW

**Comment by koppor** [comment]
This is a good start. Please add screenshots.

If you are good in Java-coding, I would like to take the screenshots using a TestFX test. See https://github.com/TestFX/TestFX for details on TestFX. There should be some tests in our code.

**Comment by michizhou** [comment]
@koppor I have added some screenshots I took of the JabRef application with the push-to-external application button and dropdown menu highlighted without using TestFX. If these are not sufficient, please let me know and I can try to take the screenshots using TestFX.

**Comment by koppor** [comment]
OK. - When reading the text, I do not see which tools are there. I remember that we have preferences where the user can configure paths etc. Is this still the case? Can you check (and provide a screenshot)?

I also found following FAQ. - Can you include the text here, too? Just the answer is enough.

#### Q: JabRef does not push to vim, although I have configured the right path and server name. What is going on?

A: You have to start vim with the option `--servername` (such as `vim --servername MyVimServer`). If you get the `Unknown option argument` message, it means your version of  vim does not include the *clientserver* feature (you can check with `vim --version`). In such a case, you have to install another version of `vim`.

**Comment by michizhou** [comment]
@koppor If you're talking about the Preferences window that is reminiscent of the following window below, then that window will pop up when you select **Options -> Preferences** in the top toolbar.
Preferences_Window
Preferences_Window
Should I include these screenshots in the documentation? Or did you have something else in mind?

**Comment by Siedlerchr** [comment]
You can select the Push-App in the Preferences (External Programs) from the drop down/combobox
This then changes the icon-button in the toolbar.
See https://github.com/JabRef/jabref/issues/5375#issuecomment-547044184

I think it would be nice to illustrate it with screenshots, showing the effect and where the "Push"-Button is located.

**Comment by koppor** [comment]
Other screenshots: https://github.com/koppor/jabref/issues/367. You can just save them and re-use them.

**Comment by koppor** [comment]
Please also add a screenshot of TeXStudio which shows that the entry is pasted. Do you have a minimal latex document (and a minimal bibliography)?

There should be two screenshots. One for `\cite{entryOne}` and one for `\cite{entryOne,entryTwo}` so that it is illustrated, that first the complete `\cite` is inserted, then the cite is extended.

**Comment by michizhou** [comment]
@koppor I've added additional screenshots and expanded documentation for the push-to-external application feature section as a whole. Let me know if this will suffice.

**Comment by koppor** [comment]
Thank you for keeping working on this. Documentation writing is hard. Even for "small features"

May I ask which version you used? I remember using http://builds.jabref.org/master/. There, I did not see any drop-down menu. Did you?

Refs https://github.com/koppor/jabref/issues/367 (where I posted screenshots)

**Comment by michizhou** [comment]
@koppor Not a problem. It's my pleasure to work on this, as I am interested in its resolution and do think it has the potential to benefit many JabRef users.

As for the version of JabRef I used, my version is 4.3.1 (the latest stable release for MacOS), and I had downloaded it from Fosshub, not http://builds.jabref.org/master/. I'll download the dev build from there (version 5.0.0) and redo the screenshots if need be.

**Comment by michizhou** [comment]
@koppor If you could look over my most recent changes to the push application documentatioon whenever you are able, I would greatly appreciate it. Thank you.

File: en/ExternalFiles.md [hunk 1]
```diff
@@ -16,6 +16,32 @@ These are by default set to values that probably make sense for your operating s
...
 To change the external viewer settings, go to **Options → Preferences → External programs**.
 
+## Pushing to external viewer application
+
+JabRef allows you to push any entries in your main window to an external viewer through the push-to-external application feature. You would need to first select the entries in your entry table that you would like to export before using the feature. Once you have done so, go to the tools submenu and click on the push-to-external application button to the left of the **Generate BibTeX keys** button. By default the external viewer used to push exports is TeXstudio.
+
+On MacOS:
+![Push to External MacOS](images/Push-External-Button-MacOS.png)
+
+On Windows:
```
Comment (koppor) [inline en/ExternalFiles.md]: Could you add two spaces (`  `) after "On Windows:"? This leads to a hard line break. (Please also at "On Mac OS:")

File: en/ExternalFiles.md [hunk 2]
```diff
@@ -16,6 +16,32 @@ These are by default set to values that probably make sense for your operating s
...
+
+![Select External Application](images/During-Application-Selection.png)
+
+Once you have made your selection and click **Save**, the push-to-external application button icon will change to match that of the selected external viewer application.
+
+![New Application After Select](images/After-Application-Selection.png)
+
+When you click on the push-to-external application button, JabRef will export your selected entries to a file generated by the selected external viewer application. As an example, here is what happens when you export one entry to TexStudio.
+
+![Initial Push to External Export](images/Initial-Push-Export.png)
```
Comment (koppor) [inline en/ExternalFiles.md]: Thank you for installing the tools. The thing is, that one should have a `.tex` file opened. You can use that file: https://github.com/latextemplates/scientific-thesis-template/blob/master/latexhints-english.tex#L43

Adding a `cite` command to a BibTeX file leads to compilation errors.

grafik

Comment (michizhou) [reply en/ExternalFiles.md]: I don't quite understand what you want. When I export a bibliography in JabRef using push-to-external application, it gets all existing entries in the bibliography and I get the following: 

After exporting first entry to TexStudio
Screen Shot 2019-11-06 at 5 11 21 PM

After exporting additional (second) entry to TexStudio
Screen Shot 2019-11-06 at 5 14 25 PM

Should I use these screenshots instead? Or are you saying that the exported file has to be in `.tex` format (which I then show screenshots for)? If it's the latter, I'd appreciate it if you could guide me on how to get the proper `.tex` file for the exported entries via JabRef (or am I supposed to use the file you linked as an example). Please let me know.

**Comment by koppor** [comment]
Second one.

The typical workflow is: Open the bib in JabRef. manage entries there. Open
the Tex file on texstudio. Produce scientific content there. Have pdflatex
and biber do their work to produce a PDF from .tex and .bib. -- In other
words: A bib file is typically not opened in texstudio (because JabRef is
used to edit the bib file. JabRef stores the bib file directly. No need for
an export)


michizhou <notifications@github.com> schrieb am Mi., 6. Nov. 2019, 23:17:

**Comment by michizhou** [comment]
@koppor I apologize for the relatively long time it took for me to push the latest changes, but I have replaced the erroneous TeXStudio export screenshots with proper ones this time. I also updated the descriptions related to those screenshots accordingly and took care of formatting. Please let me know if further changes are needed before final merging.

**Comment by koppor** [comment]
Thank you for the follow up. One Nitpick: The icon in the toolbar does not match the text. The user changed to "TeXStudio". Could you provide an addtional screenshot with texstudio icon circled. Maybe first emacs (as it exists) and then texstudio (new)

grafik


Can you move your additoins to a new file `push-to-applications.md`? I think, the text is better placed in a separate md file than in "External files", where DOIs and external links are handled.

**Comment by michizhou** [comment]
@koppor I replaced the incorrect Preferences window screenshot with a new one that indicates a user selection for Emacs as their new external viewer application. I also moved all my changes to a new separate file as per your suggestion. I would appreciate your feedback. Thanks!

**Comment by koppor** [comment]
Looks good. TeXStudio should always be spelled like that in the text. Moreover, it is an "editor", not a "viewer". Can you please quickly change that?

**Comment by michizhou** [comment]
@koppor I have made the minor edits you pointed out. Please review them when you can.

**Comment by koppor** [comment]
Thank you for following up!

I hope, you will continue creating documentation. May I propose https://github.com/JabRef/help.jabref.org/issues/93?

**Comment by michizhou** [comment]
@koppor It was my pleasure. I will definitely consider contributing more documentation in the future. Thank you again for all your help and support for this issue!

**Comment by Siedlerchr** [comment]
@michizhou  Thanks again for your valuable contribution and! This is very much appreciated and will help the users of JabRef very much!

**Comment by michizhou** [comment]
@Siedlerchr I forgot to ask, but since I have made five PRs (all of which were made in October) that have since been merged, would I be eligible for the JabRef Hacktoberfest swag? If so, I'd greatly appreciate it if you could let me know how to provide my shipping address to claim the swag as well as any other personal information I would need to provide. Thank you.

**Comment by koppor** [comment]
@michizhou Thank you for reaching us out. Also with your help the new user documentation page at https://docs.jabref.org/ looks very nice.

We just published a blog post reflecting on #hacktoberfest. -- Find my email address on my GitHub profile. Drop me a short note with your shipping address. Please give us a few weeks to prepare the T-Shirts.

We are currently busy with preparing the 5.0 release. It would be very nice if you continued supporting us so that we can release faster!

**Pull request merged**